mapie/regression/utils/ccp_phi_function.py

```diff
@@ -0,0 +1,974 @@
+                    f"The functions at index ({', '.join(map(str, inds))}) "
+                    + "of the 'functions' argument, has an unknown optional "
+                    + f"argument '{param}'.\n"
+                )
+            warnings.warn(
+                "WARNING: Unknown optional arguments.\n"
+                + warn_msg +
+                "The only recognized arguments are : 'X', 'y_pred' and 'z'. "
+                "The other optional arguments will act as parameters, "
+                "as it is always their default value which will be used."
+            )
+        if len(error_ind) > 0:
+            error_msg = ""
+            for param, inds in error_ind.items():
+                error_msg += (
+                    f"The functions at index ({', '.join(map(str, inds))}) "
+                    + "of the 'functions' argument, has an unknown required "
+                    + f"argument '{param}'.\n"
+                )
+            raise ValueError(
+                "Forbidden required argument.\n"
+                f"{error_msg}"
+                "The only allowed required argument are : 'X', "
+                "'y_pred' and 'z'.\n"
+                "Note: You can use optional arguments if you want "
+                "to. They will act as parameters, as it is always "
+                "their default value which will be used."
+            )
+        return cast(NDArray, self.functions)
+
+    @abstractmethod
+    def fit(
+        self,
+        X: ArrayLike,
+    ) -> None:
+        """
+        Fit function : Set all the necessary attributes to be able to transform
+        ``(X, y_pred, z)`` into the expected transformation.
+
+        It should set all the attributes of ``fit_attributes``
+
+        Parameters
+        ----------
+        X : Optional[ArrayLike]
+            Samples
+        """
+
+    def transform(
+        self,
+        X: Optional[ArrayLike] = None,
+        y_pred: Optional[ArrayLike] = None,
+        z: Optional[ArrayLike] = None,
+        disable_marginal_guarantee: bool = False,
+    ) -> NDArray:
+        check_is_fitted(self, self.fit_attributes)
+        self._check_transform_parameters()
+
+        params_mapping = {"X": X, "y_pred": y_pred, "z": z}
+        res = []
+
+        funct_list = list(self.functions_)
+        if not disable_marginal_guarantee and self.marginal_guarantee:
+            funct_list.append(lambda X: np.ones((len(X), 1)))
+
+        for f in funct_list:
+            params = inspect.signature(f).parameters
+
+            used_params = {
+                p: params_mapping[p] for p in params
+                if p in params_mapping and params_mapping[p] is not None
+            }
+
+            if isinstance(f, PhiFunction) and not f.normalized:
+                used_params["disable_marginal_guarantee"] = True
+
+            res.append(np.array(f(**used_params), dtype=float))
+
+            if len(res[-1].shape) == 1:
+                res[-1] = np.expand_dims(res[-1], axis=1)
```

Comment:
Can be externalised in a utils function.

Reply:
Done, in ``utils.concatenate_functions``